File: 프로그래머스/LV0/코딩테스트 입문/특이한 정렬/특이한 정렬.py
def solution(numlist, n):
    return sorted(numlist, key=lambda x : (abs(x-n), -x))


# import numpy as np
# def solution(numlist, n):
#     answer = []
#     distances = abs(np.array(numlist)-n)

#     dict1 = dict()

#     for i in range(len(numlist)):
#         if distances[i] not in dict1:
#             dict1[distances[i]] = [numlist[i]]
#         else:
#             dict1[distances[i]].append(numlist[i])

#     lst = list(dict1.keys())
#     lst.sort()

#     for k in lst:
#         answer += sorted(dict1[k], reverse=True)
#     return answer

File: 프로그래머스/LV0/코딩테스트 입문/특이한 정렬/특이한 정렬.c
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>

struct nd
{
    int number;
    int distance;
}typedef ND;

int compare(const ND* arg1, const ND* arg2)
{
    if (arg1->distance != arg2->distance)
        return (arg1->distance - arg2->distance);

    return (arg2->number - arg1->number);
}

// numlist_len은 배열 numlist의 길이입니다.
int* solution(int numlist[], size_t numlist_len, int n) {
    // return 값은 malloc 등 동적 할당을 사용해주세요. 할당 길이는 상황에 맞게 변경해주세요.
    int* answer = (int*)malloc(sizeof(int) * numlist_len);
    int* distances = (int*)malloc(sizeof(int) * numlist_len);

    ND* nd = (ND*)malloc(sizeof(ND) * numlist_len);

    for (int i = 0; i < numlist_len; i++)
    {
        distances[i] = numlist[i] - n;
        if (distances[i] < 0)
            distances[i] *= -1;

        nd[i].number = numlist[i];
        nd[i].distance = distances[i];
    }
    qsort((void*)nd, numlist_len, sizeof(ND), compare);
   
    for (int i = 0; i < numlist_len; i++)
    {
        answer[i] = nd[i].number;
    }
    return answer;
}